include/deal.II/base/linear_index_iterator.h

```diff
@@ -0,0 +1,499 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_linear_index_iterator_h
+#define dealii_linear_index_iterator_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+/**
+ * Many classes in deal.II, such as FullMatrix, TransposeTable, and
+ * SparseMatrix, store their data in contiguous buffers (though the
+ * <em>interpretation</em> of what the elements of these buffers represent
+ * can, of course, be complex). For example, FullMatrix and TransposeTable
+ * store their data in row major and column major order respectively, whereas
+ * for SparseMatrix the mapping from buffer location to matrix entry
+ * $\mathbf{A}(i, j)$ is more complicated. In any case, however, the
+ * contiguous arrangements of elements enables random access iteration.
+
+ * LinearIndexIterator provides most of the functionality needed to write
+ * iterators for these classes. LinearIndexIterator is essentially a
+ * simplified version of <code>boost::iterator_facade</code> that assumes
+ * <code>AccessorType</code> provides certain members (documented below) that
+ * completely describe the state of the iterator. The intended use of this
+ * class is for containers to define their own accessor classes and then use
+ * the curiously recurring template pattern (CRTP) technique to define their
+ * iterators, e.g.,
+ *
+ * @code
+ * // forward declaration for pointer definition
+ * template <typename T>
+ * class Container;
```

Comment:
can you say somewhere that `Container<T>` is in fact the container into which we want to have iterators?

Is it important that it actually has a template argument?